### PR TITLE
Validate operation energy estimators against source models

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- add a reproducible benchmark comparing SpikingJelly energy scores with pinned SpikeSim and NeuroMC author code and Lemaire paper equations
- publish case-level score pairs, scale-free correlation metrics, a scatter plot, and concise limitations in the Chinese and English tutorials
- clarify each estimator scope and correct the pinned SpikeSim source URL

The benchmark evaluates relative trends only. It does not treat correlation as evidence of absolute energy accuracy or hardware fidelity.

## Validation

- `186 passed, 5 skipped` across the related operation-counter and energy-estimator tests
- `uv format --check` and Ruff checks passed
- Change Log generation check passed
- Sphinx HTML build succeeded
- branch rebased onto `origin/master` at `b6a9c459`; no merge conflicts remain

## Summary by CodeRabbit

* **Documentation**
  * Expanded operation-counter tutorials with runtime modes, diagnostics, energy models, assumptions, limitations, and usage guidance in English and Chinese.
  * Added reproducible validation methodology, benchmark results, and instructions for interpreting energy estimates.
  * Updated SpikeSim source references and documented validation findings in the changelog.

* **Benchmarks**
  * Added validation coverage for Lemaire, SpikeSim, and NeuroMC forward-energy estimators using pinned references and scale-free ranking metrics.
  * Added tools for reproducing benchmark runs, comparing estimates, and generating result summaries.

### CODE REVIEW

File: benchmark/energy_model_validation.py [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+                    "torch_version": torch.__version__,
+                }
+            )
+
+
+def _plot(scores: list[Score], metrics: dict[str, dict[str, float]]) -> None:
+    models = list(metrics)
+    colors = ("#0072B2", "#D55E00", "#009E73")
+    fig, (scatter, bars) = plt.subplots(1, 2, figsize=(10.5, 4.2))
+    for model, color in zip(models, colors, strict=True):
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:606]: Not changing this. The benchmark intentionally fixes the externally validated set to SpikeSim dense, Lemaire, and NeuroMC. `strict=True` makes a model-set change fail until its cases, colors, metrics, figure, and tutorial results are updated together.

File: benchmark/energy_model_validation.py [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+def _run_author_adapter(code: str, payload: Any, cwd: Path) -> Any:
+    result = subprocess.run(
+        [sys.executable, "-c", code],
+        cwd=cwd,
+        env={**os.environ, "MPLBACKEND": "Agg"},
+        input=json.dumps(payload),
+        text=True,
+        capture_output=True,
+        check=True,
+    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:137]: Fixed in 9d6163f5 by applying a 600-second timeout to each external author adapter. Verified both the forced timeout path and the pinned SpikeSim/NeuroMC adapters.

File: benchmark/energy_model_validation.py [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+        )
+    stress_ratios = [
+        score.spikingjelly_pj / score.oracle_pj
+        for score in scores
+        if not score.included_in_correlation
+    ]
+    print(
+        f"SpikeSim dynamic stress: n={len(stress_ratios)}, "
+        f"runtime/oracle={min(stress_ratios):.3f}x..{max(stress_ratios):.3f}x"
+    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:736]: Not changing this. The five excluded dynamic-stress cases are a documented benchmark invariant. If they disappear, the benchmark and tutorial results are inconsistent and should fail instead of silently reporting no stress cases.

File: benchmark/energy_model_validation.py [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+        x = torch.zeros(1, cin, dim + kernel - 1, dim + kernel - 1)
+        report = op_counter.estimate_spikesim_energy(model, x)
+        scores.append(
+            Score(
+                model="SpikeSim dense",
+                case_id=case_id,
+                workload="static topology vs executed calls",
+                oracle_pj=author[case_id],
+                spikingjelly_pj=report.energy_total_pj,
+                oracle_source="author code: ela_spikesim.py compute_energy",
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:223]: Not changing this. Direct access to the typed SpikeSim report field is the contract and already raises an explicit `AttributeError` at this exact point if renamed; `hasattr` would not fail any earlier.

File: benchmark/energy_model_validation.py [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+        )
+    stress_ratios = [
+        score.spikingjelly_pj / score.oracle_pj
+        for score in scores
+        if not score.included_in_correlation
+    ]
+    print(
+        f"SpikeSim dynamic stress: n={len(stress_ratios)}, "
+        f"runtime/oracle={min(stress_ratios):.3f}x..{max(stress_ratios):.3f}x"
+    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:736]: Same invariant as the earlier thread: the benchmark intentionally contains five documented divergence cases. Removing all of them must fail until the benchmark results and tutorial are updated together.

File: benchmark/energy_model_validation.py [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+        dynamic_model = _lemaire_model(kind, cin, cout, spatial, kernel, lif)
+        report = op_counter.estimate_lemaire_energy(dynamic_model, x)
+        scores.append(
+            Score(
+                model="Lemaire",
+                case_id=case_id,
+                workload=f"{kind} + {'LIF' if lif else 'IF'}",
+                oracle_pj=oracle,
+                spikingjelly_pj=report.total_pj,
+                oracle_source="paper equations (1)-(20); no released code",
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:377]: Not changing this. Direct access to the typed Lemaire report field is the contract and already raises an explicit `AttributeError` at this exact point if renamed; an adjacent `hasattr` assertion adds no diagnostic value.

File: benchmark/energy_model_validation.py [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+        x = torch.zeros(1, cin, oy + fy - 1, ox + fx - 1)
+        report = op_counter.estimate_neuromc_runtime_energy(model, x)
+        scores.append(
+            Score(
+                model="NeuroMC",
+                case_id=f"n{item['layer']:02d}",
+                workload="official S-ResNet18 FE fragment",
+                oracle_pj=float(item["energy"]),
+                spikingjelly_pj=report.energy_total_pj,
+                oracle_source="author ZigZag fixed workload/mapping/cost model",
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:493]: Not changing this. Direct access to the typed NeuroMC report field is the contract and already raises an explicit `AttributeError` at this exact point if renamed; `hasattr` would not fail any earlier.

File: benchmark/energy_model_validation.py [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+            values.append(float(value))
+    if not values:
+        return math.nan, math.nan
+    return tuple(float(v) for v in np.percentile(values, [2.5, 97.5]))
+
+
+def _metrics(scores: list[Score]) -> dict[str, float]:
+    oracle = np.asarray([score.oracle_pj for score in scores], dtype=float)
+    dynamic = np.asarray([score.spikingjelly_pj for score in scores], dtype=float)
+    log_ratio = np.log(dynamic / oracle)
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:530]: Fixed in fa5d85bc. `_metrics` now rejects empty inputs and non-finite or non-positive scores before any logarithm or correlation calculation. Invalid external scores fail clearly instead of producing NaN output.

File: benchmark/energy_model_validation.py [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+    colors = ("#0072B2", "#D55E00", "#009E73")
+    fig, (scatter, bars) = plt.subplots(1, 2, figsize=(10.5, 4.2))
+    for model, color in zip(models, colors, strict=True):
+        group = [score for score in scores if score.model == model]
+        included = [score for score in group if score.included_in_correlation]
+        excluded = [score for score in group if not score.included_in_correlation]
+        oracle_scale = np.exp(np.mean(np.log([score.oracle_pj for score in included])))
+        dynamic_scale = np.exp(
+            np.mean(np.log([score.spikingjelly_pj for score in included]))
+        )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/energy_model_validation.py:613]: Not adding a second validation path here. `main()` computes `_metrics` for every plotted model first, and fa5d85bc now rejects empty or invalid scores centrally. Skipping a bad model in `_plot` would hide invalid benchmark data.
